### PR TITLE
fix dashboard save failure after deleting tabs

### LINKED ISSUE #30923: Saving a dashboard fails after creating and deleting multiple tabs then adding a card
### Describe the bug

After creating and deleting multiple tabs, and adding a card, saving a dashboard can fail because the ID of the tab isn't the same as the ID of the card. Purely a frontend bug.

### To Reproduce

Steps in Loom: https://www.loom.com/share/7795dc13da184d1fa550c0f16764c42e

Request to PUT /api/dashboard/:id/cards:
<details><summary>Payload</summary>

```
{
  "cards": [
    {
      "id": -24,
      "card_id": null,
      "dashboard_tab_id": -76,
      "row": 0,
      "col": 0,
      "size_x": 4,
      "size_y": 4,
      "series": [],
      "visualization_settings": {
        "virtual_card": {
          "name": null,
          "display": "text",
          "visualization_settings": {},
          "dataset_query": {},
          "archived": false
        },
        "text": "text"
      },
      "parameter_mappings": []
    }
  ],
  "ordered_tabs": [
    {
      "id": -75,
      "name": "Page 1"
    }
  ]
}
```
</details>

<details><summary>Response including truncated stacktrace</summary>

```
{
    "via": [
        {
            "type": "clojure.lang.ExceptionInfo",
            "message": "ERROR: insert or update on table \"report_dashboardcard\" violates foreign key constraint \"fk_report_dashboardcard_ref_dashboard_tab_id\"\n  Detail: Key (dashboard_tab_id)=(-76) is not present in table \"dashboard_tab\".",
            "data": {
                "toucan2/context-trace": [
                    [
                        "execute SQL with class com.mchange.v2.c3p0.impl.NewProxyConnection",
                        {
                            "toucan2.jdbc.query/sql-args": [
                                "INSERT INTO \"report_dashboardcard\" (\"size_x\", \"dashboard_tab_id\", \"updated_at\", \"col\", \"parameter_mappings\", \"card_id\", \"entity_id\", \"visualization_settings\", \"size_y\", \"dashboard_id\", \"created_at\", \"row\") VALUES (?, ?, NOW(), ?, ?, NULL, ?, ?, ?, ?, NOW(), ?)",
                                4,
                                -76,
                                0,
                                "[]",
                                "1LmHNvdVk2lFe5d6L6jx4",
                                "{\"virtual_card\":{\"name\":null,\"display\":\"text\",\"visualization_settings\":{},\"dataset_query\":{},\"archived\":false},\"text\":\"text\"}",
                                4,
                                25,
                                0
                            ]
                        }
                    ],
                    [
                        "resolve connection",
                        {
                            "toucan2.connection/connectable": "class org.postgresql.jdbc.PgConnection"
                        }
                    ],
                    [
                        "resolve connection",
                        {
                            "toucan2.connection/connectable": null
                        }
                    ],
                    {
                        "toucan2.pipeline/rf": "clojure.core$map$fn__5931$fn__5932@1b07b8d4"
                    },
                    [
                        "with compiled query",
                        {
                            "toucan2.pipeline/compiled-query": [
                                "INSERT INTO \"report_dashboardcard\" (\"size_x\", \"dashboard_tab_id\", \"updated_at\", \"col\", \"parameter_mappings\", \"card_id\", \"entity_id\", \"visualization_settings\", \"size_y\", \"dashboard_id\", \"created_at\", \"row\") VALUES (?, ?, NOW(), ?, ?, NULL, ?, ?, ?, ?, NOW(), ?)",
                                4,
                                -76,
                                0,
                                "[]",
                                "1LmHNvdVk2lFe5d6L6jx4",
                                "{\"virtual_card\":{\"name\":null,\"display\":\"text\",\"visualization_settings\":{},\"dataset_query\":{},\"archived\":false},\"text\":\"text\"}",
                                4,
                                25,
                                0
                            ]
                        }
                    ],
                    [
                        "with built query",
                        {
                            "toucan2.pipeline/built-query": {
                                "insert-into": [
                                    "report_dashboardcard"
                                ],
                                "values": [
                                    [
                                        "toucan2.instance/instance",
                                        "model/DashboardCard",
                                        {
                                            "size_x": 4,
                                            "dashboard_tab_id": -76,
                                            "updated_at": [
                                                "metabase.util.honey-sql-2/typed",
                                                "%now",
                                                {
                                                    "metabase.util.honeysql-extensions/database-type": "timestamptz"
                                                }
                                            ],
                                            "col": 0,
                                            "parameter_mappings": "[]",
                                            "card_id": null,
                                            "entity_id": "1LmHNvdVk2lFe5d6L6jx4",
                                            "visualization_settings": "{\"virtual_card\":{\"name\":null,\"display\":\"text\",\"visualization_settings\":{},\"dataset_query\":{},\"archived\":false},\"text\":\"text\"}",
                                            "size_y": 4,
                                            "dashboard_id": 25,
                                            "created_at": [
                                                "metabase.util.honey-sql-2/typed",
                                                "%now",
                                                {
                                                    "metabase.util.honeysql-extensions/database-type": "timestamptz"
                                                }
                                            ],
                                            "row": 0
                                        }
                                    ]
                                ]
                            }
                        }
                    ],
                    [
                        "resolve connection",
                        {
                            "toucan2.connection/connectable": "class org.postgresql.jdbc.PgConnection"
                        }
                    ],
                   ....
    "cause": "ERROR: insert or update on table \"report_dashboardcard\" violates foreign key constraint \"fk_report_dashboardcard_ref_dashboard_tab_id\"\n  Detail: Key (dashboard_tab_id)=(-76) is not present in table \"dashboard_tab\".",
    "message": "ERROR: insert or update on table \"report_dashboardcard\" violates foreign key constraint \"fk_report_dashboardcard_ref_dashboard_tab_id\"\n  Detail: Key (dashboard_tab_id)=(-76) is not present in table \"dashboard_tab\"."
}
```

### Expected behavior

dashboard should save correctly

### Logs

_No response_

### Information about your Metabase installation

```JSON
master
```

### Severity

high

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by EmmadUsmani**
Closes https://github.com/metabase/metabase/issues/30923

### Description

When we added the undo toast notification for deleting tabs in https://github.com/metabase/metabase/pull/30497, a regression was introduced where saving a dashboard could fail after deleting tabs.

This PR fixes this by updating the logic for tab deletion to understand that tabs now follow a tombstone deletion model, thus we ignore tabs that have an `isRemoved: true` attribute.

### How to verify

1. New dashboard
2. Add two new tabs
3. Delete `Page 2`
4. Delete `Page 3`
5. Add a card
6. Save

### Demo

https://github.com/metabase/metabase/assets/37751258/a46021cc-1b09-43b1-a2e0-681ccad91ec0

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

**Review by JesseSDevaney — CHANGES_REQUESTED** [review]
This fixes the targeted issue, but accidentally introduces a regression in un-doing tab deletions. I made a suggestion with a couple ideas for how to fix it :heart:

**Comment by JesseSDevaney** [comment]
Also, unrelated to this issue and I may be incorrect, but I think there is some unused code in the `deleteTab` case.
```typescript
builder.addCase(
      deleteTab,
      (state, { payload: { tabId, tabDeletionId } }) => {
        const { dashId, prevDash, prevTabs } = getPrevDashAndTabs(state);
        const tabToRemove = prevTabs.find(({ id }) => id === tabId);
        if (!dashId || !prevDash || !tabToRemove) {
          throw Error(
            `DELETE_TAB was dispatched but either dashId (${dashId}), prevDash (${prevDash}), or tabToRemove (${tabToRemove}) is null/undefined`,
          );
        }
        // 1. Select a different tab if needed
        const noTabsRemaining = prevTabs.length === 1; // **************************
        const deletingSelectedTab = state.selectedTabId === tabToRemove.id;

        if (noTabsRemaining) { // **************************
          state.selectedTabId = null; // **************************
        } else if (deletingSelectedTab) {
          const tabToRemoveIndex = prevTabs.findIndex(
            ({ id }) => id === tabToRemove.id,
          );
          const targetIndex = tabToRemoveIndex === 0 ? 1 : tabToRemoveIndex - 1;
          state.selectedTabId = prevTabs[targetIndex].id;
        }
        // 2. Mark the tab as removed
        tabToRemove.isRemoved = true;
        // 3. Mark dashcards on removed tab as removed
        const removedDashCardIds: DashCardId[] = [];
        prevDash.ordered_cards.forEach(id => {
          if (state.dashcards[id].dashboard_tab_id === tabToRemove.id) {
            state.dashcards[id].isRemoved = true;
            removedDashCardIds.push(id);
          }
        });
        // 4. Add deletion to history to allow undoing
        state.tabDeletions[tabDeletionId] = {
          id: tabDeletionId,
          tabId: tabToRemove.id,
          removedDashCardIds,
        };
      },
    );
```

I think the unused code is:
```typescript
const noTabsRemaining = prevTabs.length === 1;

if (noTabsRemaining) {
  state.selectedTabId = null;
```
Is there actually a point where there would be only one tab and we are able to delete it? Deleting a tab is disabled for dashboards with only one tab if I am not wrong?

File: frontend/src/metabase/dashboard/actions/tabs.ts [hunk 1]
```diff
@@ -64,7 +69,7 @@ export const initTabs = createAction(INIT_TABS);
 function getPrevDashAndTabs(state: Draft<DashboardState>) {
   const dashId = state.dashboardId;
   const prevDash = dashId ? state.dashboards[dashId] : null;
-  const prevTabs = prevDash?.ordered_tabs ?? [];
+  const prevTabs = prevDash?.ordered_tabs?.filter(t => !t.isRemoved) ?? [];
```
[the remark this answers is not in the document]
Comment (EmmadUsmani) [reply frontend/src/metabase/dashboard/actions/tabs.ts:72]: Thanks for the thorough explanation and suggestions! I'll the code to use 2

**Comment by EmmadUsmani** [comment]
> Also, unrelated to this issue and I may be incorrect, but I think there is some unused code in the deleteTab case.

@JesseSDevaney Nice catch! The original behavior was that you could delete the last tab, I forgot to remove this code when we changed it. I'll update this as well.

**Review by alxnddr — APPROVED** [review]

**Review by JesseSDevaney — APPROVED** [review]

**Pull request merged**